### PR TITLE
Core: Handles possible heap data corruption of `OAuth2Util.AuthSession#headers`

### LINKED ISSUE #10591: Possible heap data corruption of `org.apache.iceberg.rest.auth.OAuth2Util.AuthSession#headers`
### Apache Iceberg version

None

### Query engine

None

### Please describe the bug 🐞

The operation on the volatile `org.apache.iceberg.rest.auth.OAuth2Util.AuthSession#headers` is not atomic.

### PR DESCRIPTION
**Pull request by tlm365**
Closes #10591.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 1]
```diff
@@ -578,7 +579,9 @@ public Pair<Integer, TimeUnit> refresh(RESTClient client) {
                 .token(response.token())
                 .tokenType(response.issuedTokenType())
                 .build();
-        this.headers = RESTUtil.merge(headers, authHeaders(config.token()));
+        synchronized (lock) {
```
[the remark this answers is not in the document]
Comment (adutra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:582]: TBH I'm not convinced that a lock solves a real problem. In practice, this method cannot be called concurrently, therefore just dissociating reading the field and writing the field back is enough to clarify the intents here:

```java
Map<String, String> headers = this.headers;
this.headers = RESTUtil.merge(headers, authHeaders(config.token()));
```

[the remark this answers is not in the document]
Comment (tlm365) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:582]: @snazy @adutra thanks so much for reviewing
> Maybe just a simple?

@snazy I just wanna avoid using `synchronized` on class. See: https://stackoverflow.com/questions/442564/avoid-synchronizedthis-in-java

> TBH I'm not convinced that a lock solves a real problem. In practice, this method cannot be called concurrently, therefore just dissociating reading the field and writing the field back is enough to clarify the intents here:

@adutra "this method cannot be called concurrently" - could you explain why, pls? The `RESTUtil.merge(...)` function read a `volatile` variable (headers), so I think `synchronized` might make sense here.

**Comment by adutra** [comment]
@tlm365 `AuthSession.refresh()` is invoked from two places:

* `fromAccessToken()`: creates a new `AuthSession` instance each time, so no concurrent access can happen here.
* `scheduleTokenRefresh()`: cannot be called concurrently given the access patterns (it's called first on catalog initialization – which we can assume is single-threaded –, then called by itself recursively from within the refresh executor thread).

**Review by amogh-jahagirdar — COMMENTED** [review]
I agree with @adutra , I don't think a lock is necessary here. You can just read the volatile value into a temporary variable and use that temporary variable when merging and assign that to the field.

**Comment by tlm365** [comment]
@adutra @amogh-jahagirdar thanks for reviewing, I've updated it.

**Review by amogh-jahagirdar — APPROVED** [review]
Thanks @tlm365 I will merge when checks pass.

**Pull request merged**